internal/directives/json_updater.go

```diff
@@ -0,0 +1,150 @@
+}
+
+// jsonUpdater is an implementation of the PromotionStepRunner interface that
+// updates the values of specified keys in a JSON file.
+type jsonUpdater struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newJSONUpdater returns an implementation of the PromotionStepRunner interface
+// that updates the values of specified keys in a JSON file.
+func newJSONUpdater() PromotionStepRunner {
+	r := &jsonUpdater{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the PromotionStepRunner interface.
+func (j *jsonUpdater) Name() string {
+	return "json-update"
+}
+
+// RunPromotionStep implements the PromotionStepRunner interface.
+func (j *jsonUpdater) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	failure := PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}
+
+	if err := j.validate(stepCtx.Config); err != nil {
+		return failure, err
+	}
+
+	// Convert the configuration into a typed struct
+	cfg, err := ConfigToStruct[JSONUpdateConfig](stepCtx.Config)
+	if err != nil {
+		return failure, fmt.Errorf("could not convert config into %s config: %w", j.Name(), err)
+	}
+
+	return j.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates jsonUpdater configuration against a JSON schema.
+func (j *jsonUpdater) validate(cfg Config) error {
+	return validate(j.schemaLoader, gojsonschema.NewGoLoader(cfg), j.Name())
+}
+
+func (j *jsonUpdater) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg JSONUpdateConfig,
+) (PromotionStepResult, error) {
+	updates := make(map[string]any, len(cfg.Updates))
+	for _, update := range cfg.Updates {
+		var value any
+		if update.Value != nil {
+			value = update.Value
+		}
+		updates[update.Key] = value
+	}
+
+	result := PromotionStepResult{Status: kargoapi.PromotionPhaseSucceeded}
+	if len(updates) > 0 {
+		if err := j.updateFile(stepCtx.WorkDir, cfg.Path, updates); err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("JSON file update failed: %w", err)
+		}
+
+		if commitMsg := j.generateCommitMessage(cfg.Path, updates); commitMsg != "" {
+			result.Output = map[string]any{
+				"commitMessage": commitMsg,
+			}
+		}
+	}
+	return result, nil
+}
+
+func (j *jsonUpdater) updateFile(workDir string, path string, changes map[string]any) error {
+	absFilePath, err := securejoin.SecureJoin(workDir, path)
+	if err != nil {
+		return fmt.Errorf("error joining path %q: %w", path, err)
+	}
+
+	fileContent, err := os.ReadFile(absFilePath)
+	if err != nil {
+		return fmt.Errorf("error reading JSON file %q: %w", absFilePath, err)
+	}
+
+	for key, value := range changes {
+		updatedContent, setErr := sjson.Set(string(fileContent), key, value)
+		if setErr != nil {
+			return fmt.Errorf("error setting key %q in JSON file: %w", key, setErr)
+		}
+		fileContent = []byte(updatedContent)
+	}
+
+	err = os.WriteFile(absFilePath, fileContent, 0o600)
```

Comment:
Nit: This notation is more common.

```suggestion
	err = os.WriteFile(absFilePath, fileContent, 0600)
```